### PR TITLE
[v6] plumbing: format/index, Improve v4 entry name validation

### PR DESCRIPTION
**Pull request by pjbgf**
Updates Git index (`plumbing/format/index`) encoding/decoding logic to better match on-disk format edge cases (long entry names and V4 name prefix-compression), and adds regression tests around these behaviors.

Relates to #1868.

### CODE REVIEW

File: plumbing/format/index/decoder.go [hunk 1]
```diff
@@ -146,33 +149,55 @@ func (d *Decoder) readEntry(idx *Index) (*Entry, error) {
...
+// the 12-bit field and the real length is found by scanning for the NUL
+// terminator — matching C Git's strlen(name) fallback in create_from_disk.
+func (d *Decoder) doReadEntryName(nameLen uint16) (string, int, error) {
+	if nameLen == nameMask {
+		name, err := binary.ReadUntil(d.r, '\x00')
+		if err != nil {
+			return "", 0, err
+		}
+		return string(name), len(name) + 1, nil // +1 for the consumed NUL delimiter
 	}
```
[the remark this answers is not in the document]
Comment (pjbgf) [reply plumbing/format/index/decoder.go:196]: The performance impact seems rather minimal, but a follow-up PR will take care of additional performance optimisations.

File: plumbing/format/index/decoder.go [hunk 2]
```diff
@@ -183,7 +208,14 @@ func (d *Decoder) doReadEntryNameV4() (string, error) {
 
 	var base string
 	if d.lastEntry != nil {
+		if l < 0 || int(l) > len(d.lastEntry.Name) {
+			return "", fmt.Errorf("%w: invalid V4 entry name strip length %d (previous name length: %d)",
+				ErrMalformedIndexFile, l, len(d.lastEntry.Name))
+		}
 		base = d.lastEntry.Name[:len(d.lastEntry.Name)-int(l)]
+	} else if l > 0 {
```
[the remark this answers is not in the document]
Comment (pjbgf) [reply plumbing/format/index/decoder.go:216]: This would not panic as the `int(l)` operation would be consistent across both lines `211` and `215`. The worse case scenario here would be an incorrect base being returned.
